File: elixir/day1/lib/part2.ex
defmodule Part2 do
  def find_numbers_or_digits(line) do
    ~r/\d|(one|two|three|four|five|six|seven|eight|nine)/
    |> Regex.scan(line)
    |> List.flatten()
  end

  def number(word) do
    case word do
      "one" ->
        1

      "two" ->
        2

      "three" ->
        3

      "four" ->
        4

      "five" ->
        5

      "six" ->
        6

      "seven" ->
        7

      "eight" ->
        8

      "nine" ->
        9

      _ ->
        case Integer.parse(word) do
          {num, _} -> num
          _ -> 0
        end
    end
  end

  def part2(input) do
    input
    |> String.split("\n", trim: true)
    |> Enum.reduce(0, fn line, acc ->
      numbers =
        line
        |> find_numbers_or_digits()
        |> Enum.map(&number/1)

      acc + Enum.at(numbers, 0) * 10 + Enum.at(numbers, -1)
    end)
  end
end

File.read!("input.txt")
|> Part2.part2()
|> IO.inspect(label: "part2")

File: ruby/day1/part2.rb
# frozen_string_literal: true

def find_word_or_digit(line)
  numbers = {
    'one' => 1,
    'two' => 2,
    'three' => 3,
    'four' => 4,
    'five' => 5,
    'six' => 6,
    'seven' => 7,
    'eight' => 8,
    'nine' => 9
  }

  numbers.each do |word, number|
    return number if line.start_with?(word)
  end

  line[0].to_i if line[0].match?(/\d/)
end

def process_line(line)
  first, last = nil
  (0..line.length - 1).each do |index|
    first = find_word_or_digit(line[index..]) if first.nil?
    last = find_word_or_digit(line[index..]) || last
  end

  if first && last
    (first * 10) + last
  else
    0
  end
end

def part2(input)
  input.split("\n").sum { |line| process_line(line) }
end

if __FILE__ == $PROGRAM_NAME
  input = File.read('input.txt')

  p part2(input)
end
